keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,432 @@
+
+2. The CSI migration initiative (planned to reach beta state in v1.16) aims to deprecate the code associated with several in-tree storage plugins and pave the path for the ultimate removal of that code from Kubernetes core in favor of CSI plugins that implement the same functionality. Windows workloads need to be aligned with the CSI migration effort and cannot depend on environment specific in-tree plugins to satisfy persistent storage needs.
+
+### Goals
+
+1. Support all CSI Node Plugin operations: NodeStageVolume/NodeUnstageVolume, NodePublishVolume/NodeUnPublishVolume, NodeExpandVolume, NodeGetVolumeStats, NodeGetCapabilities and NodeGetInfo on Windows nodes.
+
+2. Support CSI plugins associated with a wide variety of external storage scenarios: remote storage over iSCSI and SMB as well as directly attached storage scenarios in cloud environments.
+
+3. Ability to distribute CSI Node Plugins targeting Windows nodes as containers that can be deployed using DaemonSets in a Kubernetes cluster comprising of Windows nodes.
+
+### Non-Goals
+
+1. Support CSI Controller Plugin operations from Windows nodes: This may be considered in the future but not an immediate priority. Note that this does not require support for privileged operations on a Windows node as required by CSI Node Plugins and thus orthogonal to this KEP around CSI Node Plugins for Windows. If all the worker nodes in the cluster are Windows nodes and Linux master nodes have scheduling disabled then CSI controller plugins cannot be scheduled for now.
+
+2. Support privileged operations from Windows containers beyond CSI Node Plugins: This KEP introduces a host based "privileged proxy" process that may be used for executing privileged operations on the host on behalf of a Windows container. While a similar mechanism may be used for other use cases like containerized CNI plugins (for executing HNS operations), we leave that for a separate KEP. Scoping down the set of actions allowed by the API exposed by by the privileged proxy process to a minimal set simplifies multiple versions of the API as well as reduces the scope for abuse.
+
+## Proposal
+
+In this KEP, we propose the following enhancements in pre-existing components to support CSI Node Plugins on Windows nodes:
+1. Ability to handle Windows file paths in the Kubelet plugin watcher for domain sockets on Windows nodes.
+2. Refactor code in the CSI Node Driver Registrar so that it can be compiled for Windows.
+3. Build official CSI Node Driver Registrar container images based on Windows base images.
+4. Refactor code in existing CSI Node Plugins to support Windows. All privileged operations will need to be driven through an API exposed by a "privileged proxy" binary described below.
```

Comment:
should we leave this step to individual plugin owners? Maybe producing a guide for driver developers to follow is more scalable